### DISCUSSION TITLE
Rust integration

### DISCUSSION THREAD

**Discussion by snth** [cunnane/xloil#39] [Announcements]
I have some ideas for integrations with xlOil which would be a lot easier with Rust. I could go the pyIodide route but directly from C++ would be better.

I haven't used C++ in 25 years so I don't know anything about that side. Would be great if we could collaborate on this.

**Comment by cunnane** [comment]
Sounds intriguing, you're looking for Rust<>Excel interop or something else?

Would this conversation be better placed in the Discussions section rather?

**Comment by ayemjay** [comment]
Very much interested in this. 

Would you think to:
1. develop a C API with Rust FFI bindings?
2. wrap around existing C++ API?
3. re-implement code base in Rust?

Least "combined" effort is `2.` as only require development in Rust but unsure how permissive bindgen in Rust is with C++ APIs.

Would there be appetite for `1.`? my C/C++ skills are lacking but could contribute on the Rust side.

**Comment by ayemjay** [comment]
As for Rust<>Excel interop following crates exist:

- https://github.com/MarcusRainbow/xladd
- https://github.com/ronniec95/xladd

These provide support for developing XLLs with UDFs but unfortunately do not provide RTD or async (actually yes but less battle tested).

**Comment by cunnane** [comment]
Sorry, it seems I don't get email updates when someone posts in a discussion!

Of your 3 options, I think (3) is the least feasible since Excel's interfaces are in C/C++.  I don't know rust well, but some google searching suggested (2) may also be difficult (at least there's no mainstream support), so I started on (1).  I think a C interface could be created semi-automatically with SWIG but the C wrappers aren't part of the main branch, so could be buggy especially with modern C++.  I started to manually create C wrappers, but it's going to be a long road!  Would your priority be to get RTD support?  (btw I don't see any RTD support in *xladd* which you linked to above)
